src/parser/sparqlParser/SparqlQleverVisitor.cpp

```diff
@@ -466,39 +444,77 @@ std::pair<vector<SparqlTripleSimple>, ParsedQuery::GraphPattern> Visitor::visit(
         AD_CORRECTNESS_CHECK(triple.p_.isVariable() || triple.p_.isIri());
         return SparqlTriple::fromSimple(triple);
       };
+  std::vector<Variable> visibleVariablesSoFar = std::move(visibleVariables_);
+  visibleVariables_.clear();
   GraphPattern pattern;
+  auto triples = visit(ctx->quadPattern());
   pattern._graphPatterns.emplace_back(BasicGraphPattern{
       ad_utility::transform(triples, transformAndRegisterTriple)});
+  parsedQuery_._rootGraphPattern = std::move(pattern);
+  auto visibleVariablesWhereClause =
+      std::exchange(visibleVariables_, std::move(visibleVariablesSoFar));
+  parsedQuery_.registerVariablesVisibleInQueryBody(visibleVariablesWhereClause);
+  // The query body and template are identical. Variables will always be visible
+  // - no need to check that.
+  parsedQuery_._clause =
+      parsedQuery::UpdateClause{GraphUpdate{{}, std::move(triples)}};
 
-  return {std::move(triples), std::move(pattern)};
+  return parsedQuery_;
 }
 
 // ____________________________________________________________________________________
 ParsedQuery Visitor::visit(Parser::ModifyContext* ctx) {
-  if (ctx->iri()) {
-    reportNotSupported(ctx->iri(), "Named graphs are");
-  }
-  if (!ctx->usingClause().empty()) {
-    reportNotSupported(ctx->usingClause(0),
-                       "USING inside an DELETE or INSERT is");
-  }
-
-  parsedQuery_._rootGraphPattern = visit(ctx->groupGraphPattern());
-
-  parsedQuery_._clause = parsedQuery::UpdateClause();
-  visitIf(&parsedQuery_.updateClause().toInsert_, ctx->insertClause());
-  visitIf(&parsedQuery_.updateClause().toDelete_, ctx->deleteClause());
+  auto isVisibleIfVariable = [this](const TripleComponent& component) {
+    if (component.isVariable()) {
+      return std::ranges::find(parsedQuery_.getVisibleVariables(),
+                               component.getVariable()) !=
+             parsedQuery_.getVisibleVariables().end();
+    } else {
+      return true;
+    }
+  };
+  auto checkTriples =
+      [&isVisibleIfVariable,
+       &ctx](const std::vector<SparqlTripleSimpleWithGraph>& triples) {
+        for (auto& triple : triples) {
+          if (!(isVisibleIfVariable(triple.s_) &&
+                isVisibleIfVariable(triple.p_) &&
+                isVisibleIfVariable(triple.o_))) {
+            reportError(ctx,
+                        absl::StrCat("A triple contains a variable that was "
+                                     "not bound in the query body."));
+          }
+        }
+      };
```

Comment:
Is this duplicated/copy-pasted from some other type of query?
If so, remove that duplication by helper functions. Else I am going to check this some other time when I am less tired, this seems to be the most complicated rule.

Reply:
AFAIK this is not duplicated. But maybe a (currently private) of `ParsedQuery` can be used instead of `isVisibleIfVariable`.